File: app/src/main/java/com/group_finity/mascot/behavior/BehaviorBuilder.java
package com.group_finity.mascot.behavior;

import com.group_finity.mascot.action.Action;
import com.group_finity.mascot.action.SequenceAction;
import com.group_finity.mascot.config.xml.XmlBehavior;
import com.group_finity.mascot.config.xml.XmlBehaviors;
import com.group_finity.mascot.config.xml.XmlActionReference;
import com.group_finity.mascot.trigger.Trigger;
import jakarta.xml.bind.JAXBContext;
import jakarta.xml.bind.JAXBException;
import jakarta.xml.bind.Unmarshaller;

import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

/**
 * Builds a list of {@link Behavior}s from an XML configuration file.
 * <p>
 * It depends on a pre-built map of named actions to link triggers with their resulting actions.
 * <p>
 * Example XML:
 * <pre>{@code
 * <behaviors>
 *   <behavior>
 *     <condition>mascot.isGrounded</condition>
 *     <action ref="Walk" />
 *   </behavior>
 * </behaviors>
 * }</pre>
 */
public class BehaviorBuilder {

    private final Map<String, Action> actions;

    public BehaviorBuilder(Map<String, Action> actions) {
        this.actions = actions;
    }

    public List<Behavior> build(Path behaviorsPath) {
        try {
            JAXBContext context = JAXBContext.newInstance(XmlBehaviors.class);
            Unmarshaller unmarshaller = context.createUnmarshaller();
            XmlBehaviors xmlBehaviors = (XmlBehaviors) unmarshaller.unmarshal(behaviorsPath.toFile());

            if (xmlBehaviors.getBehaviors() == null) {
                return Collections.emptyList();
            }

            List<Behavior> builtBehaviors = new ArrayList<>();
            for (XmlBehavior xmlBehavior : xmlBehaviors.getBehaviors()) {
                // 実行するActionを解決または生成します。
                Action action = createActionForBehavior(xmlBehavior);

                if (action != null) {
                    // Behaviorクラスを直接生成し、リストに追加します。
                    String name = (xmlBehavior.getName() != null) ? xmlBehavior.getName() : "Behavior";
                    boolean hidden = xmlBehavior.isHidden();
                    int frequency = (xmlBehavior.getFrequency() != null) ? xmlBehavior.getFrequency() : 1;
                    builtBehaviors.add(new Behavior(name, action, xmlBehavior.getCondition(), hidden, frequency));
                } else {
                    System.err.println("No valid action found for condition: " + xmlBehavior.getCondition());
                }
            }
            return Collections.unmodifiableList(builtBehaviors);

        } catch (JAXBException e) {
            System.err.println("Failed to parse behaviors.xml: " + behaviorsPath);
            e.printStackTrace();
            return Collections.emptyList();
        }
    }

    private Action createActionForBehavior(XmlBehavior xmlBehavior) {
        List<XmlActionReference> references = xmlBehavior.getActionReferences();
        if (references == null || references.isEmpty()) {
            return null;
        }

        if (references.size() == 1) {
            // ActionReferenceが1つの場合は、対応するActionをそのまま返します。
            String actionName = references.get(0).getName();
            Action action = actions.get(actionName);
            if (action == null) {
                System.err.println("Action not found for behavior: " + actionName);
            }
            return action;
        } else {
            // ActionReferenceが複数の場合は、それらを順に実行するSequenceActionを動的に生成します。
            List<Action> sequence = new ArrayList<>();
            for (XmlActionReference ref : references) {
                Action referencedAction = actions.get(ref.getName());
                if (referencedAction != null) {
                    sequence.add(referencedAction);
                } else {
                    System.err.println("ActionReference not found: " + ref.getName() + " in behavior with condition: " + xmlBehavior.getCondition());
                }
            }
            // SequenceActionは別途実装が必要です。
            return sequence.isEmpty() ? null : new SequenceAction(sequence);
        }
    }
}

File: app/src/main/kotlin/com/group_finity/mascot/behavior/BehaviorBuilder.kt
package com.group_finity.mascot.behavior

import com.group_finity.mascot.action.Action
import com.group_finity.mascot.action.SequenceAction
import com.group_finity.mascot.config.BehaviorConfig
import com.group_finity.mascot.config.XmlSecurity
import com.group_finity.mascot.config.xml.XmlBehavior
import com.group_finity.mascot.config.xml.XmlBehaviors
import jakarta.xml.bind.JAXBContext
import java.nio.file.Path
import java.util.Collections

class BehaviorBuilder(private val actions: Map<String, Action>) {

    // === XML Build Logic ===
    fun build(behaviorsPath: Path): List<Behavior> {
        return try {
            val context = JAXBContext.newInstance(XmlBehaviors::class.java)
            val unmarshaller = context.createUnmarshaller()

            val dbf = XmlSecurity.createSecureFactory()
            val db = dbf.newDocumentBuilder()
            val doc = db.parse(behaviorsPath.toFile())
            val xmlBehaviors = unmarshaller.unmarshal(doc) as XmlBehaviors

            if (xmlBehaviors.behaviors == null) {
                return emptyList()
            }

            val builtBehaviors = ArrayList<Behavior>()
            for (xmlBehavior in xmlBehaviors.behaviors) {
                val action = createActionForBehavior(xmlBehavior)

                if (action != null) {
                    val name = xmlBehavior.name ?: "Behavior"
                    val hidden = xmlBehavior.isHidden
                    val frequency = xmlBehavior.frequency ?: 1
                    builtBehaviors.add(
                            Behavior(name, action, xmlBehavior.condition, hidden, frequency)
                    )
                } else {
                    System.err.println(
                            "No valid action found for condition: " + xmlBehavior.condition
                    )
                }
            }
            Collections.unmodifiableList(builtBehaviors)
        } catch (e: Exception) {
            System.err.println("Failed to parse behaviors.xml: $behaviorsPath")
            e.printStackTrace()
            emptyList()
        }
    }

    private fun createActionForBehavior(xmlBehavior: XmlBehavior): Action? {
        val references = xmlBehavior.actionReferences
        if (references.isNullOrEmpty()) {
            return null
        }

        if (references.size == 1) {
            val actionName = references[0].name
            val action = actions[actionName]
            if (action == null) {
                System.err.println("Action not found for behavior: $actionName")
            }
            return action
        } else {
            val sequence = ArrayList<Action>()
            for (ref in references) {
                val referencedAction = actions[ref.name]
                if (referencedAction != null) {
                    sequence.add(referencedAction)
                } else {
                    System.err.println(
                            "ActionReference not found: ${ref.name} in behavior with condition: ${xmlBehavior.condition}"
                    )
                }
            }
            return if (sequence.isEmpty()) null else SequenceAction(sequence)
        }
    }

    // === YAML Build Logic ===
    fun build(behaviorConfigs: List<BehaviorConfig>?): List<Behavior> {
        if (behaviorConfigs == null) return emptyList()

        val builtBehaviors = ArrayList<Behavior>()
        for (config in behaviorConfigs) {
            val action = createActionForBehaviorConfig(config)

            if (action != null) {
                val name = config.name
                val hidden = false // YAML default
                val frequency = config.frequency
                // YAML "conditions" is a List<String>, but behavior expects a single String
                // condition.
                // We join them with "&&" or just take the first one. For now, logic match XML:
                // single string.
                // Taking first or defaulting to true.
                val conditionStr =
                        if (config.conditions.isNotEmpty()) config.conditions[0] else "true"

                builtBehaviors.add(Behavior(name, action, conditionStr, hidden, frequency))
            } else {
                System.err.println("No valid action found for behavior: ${config.name}")
            }
        }
        return Collections.unmodifiableList(builtBehaviors)
    }

    private fun createActionForBehaviorConfig(config: BehaviorConfig): Action? {
        val actionNames = config.actions
        if (actionNames.isEmpty()) {
            return null
        }

        if (actionNames.size == 1) {
            val actionName = actionNames[0]
            val action = actions[actionName]
            if (action == null) {
                System.err.println("Action not found for behavior: $actionName")
            }
            return action
        } else {
            val sequence = ArrayList<Action>()
            for (name in actionNames) {
                val referencedAction = actions[name]
                if (referencedAction != null) {
                    sequence.add(referencedAction)
                } else {
                    System.err.println(
                            "ActionReference not found: $name in behavior: ${config.name}"
                    )
                }
            }
            return if (sequence.isEmpty()) null else SequenceAction(sequence)
        }
    }
}
